server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -907,132 +914,144 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
-                        filesToSnapshot.add(snapshotFileInfo);
-                    } else {
-                        indexCommitPointFiles.add(existingFileInfo);
-                    }
                 }
 
-                snapshotStatus.moveToStarted(startTime, indexIncrementalFileCount,
-                    indexTotalNumberOfFiles, indexIncrementalSize, indexTotalFileCount);
-
-                for (BlobStoreIndexShardSnapshot.FileInfo snapshotFileInfo : filesToSnapshot) {
-                    try {
-                        snapshotFile(snapshotFileInfo, indexId, shardId, snapshotId, snapshotStatus, store);
-                    } catch (IOException e) {
-                        throw new IndexShardSnapshotFailedException(shardId, "Failed to perform snapshot (index files)", e);
-                    }
+                indexTotalFileCount += md.length();
+                indexTotalNumberOfFiles++;
+
+                if (existingFileInfo == null) {
+                    indexIncrementalFileCount++;
+                    indexIncrementalSize += md.length();
+                    // create a new FileInfo
+                    BlobStoreIndexShardSnapshot.FileInfo snapshotFileInfo =
+                        new BlobStoreIndexShardSnapshot.FileInfo(DATA_BLOB_PREFIX + UUIDs.randomBase64UUID(), md, chunkSize());
+                    indexCommitPointFiles.add(snapshotFileInfo);
+                    filesToSnapshot.add(snapshotFileInfo);
+                } else {
+                    indexCommitPointFiles.add(existingFileInfo);
                 }
-            } finally {
-                store.decRef();
             }
 
-            final IndexShardSnapshotStatus.Copy lastSnapshotStatus = snapshotStatus.moveToFinalize(snapshotIndexCommit.getGeneration());
+            snapshotStatus.moveToStarted(startTime, indexIncrementalFileCount,
+                indexTotalNumberOfFiles, indexIncrementalSize, indexTotalFileCount);
 
-            // now create and write the commit point
-            final BlobStoreIndexShardSnapshot snapshot = new BlobStoreIndexShardSnapshot(snapshotId.getName(),
-                lastSnapshotStatus.getIndexVersion(),
-                indexCommitPointFiles,
-                lastSnapshotStatus.getStartTime(),
-                threadPool.absoluteTimeInMillis() - lastSnapshotStatus.getStartTime(),
-                lastSnapshotStatus.getIncrementalFileCount(),
-                lastSnapshotStatus.getIncrementalSize()
-            );
+            assert indexIncrementalFileCount == filesToSnapshot.size();
 
-            logger.trace("[{}] [{}] writing shard snapshot file", shardId, snapshotId);
-            try {
-                indexShardSnapshotFormat.write(snapshot, shardContainer, snapshotId.getUUID());
-            } catch (IOException e) {
-                throw new IndexShardSnapshotFailedException(shardId, "Failed to write commit point", e);
-            }
+            final Runnable afterUploads = () -> {
```

Comment:
I don't really like this `Runnable`, maybe we should use `StepListener`s instead?

I'm thinking of something like:
```java
final StepListener<Void> snapshotDone = new StepListener<>();
snapshotDone.whenComplete(listener::onResponse, onFailure); // snapshot is done, call parent listener

final StepListener<Void> snapshotFinalization = new StepListener<>();
snapshotFinalization.whenComplete(v -> {
    ... what is in the runnable...
    snapshotDone.onResponse(null);
}, snapshotDone::onFailre);

if (indexIncrementalFileCount == 0) {
    snapshotFinalization.onResponse(null);
} else {
    ... use the GroupedActionListener with a wrapped snapshotFinalization...
}
```

WDYT?

Reply:
I like it :) Used a step listener the way you suggested now (with one naming change).

Reply:
This is nice :+1: